src/core/helpers/bimap.rs

```diff
@@ -0,0 +1,54 @@
+use std::{
+    borrow::Borrow,
+    collections::{btree_map::Range, BTreeMap},
+    ops::RangeBounds,
+    sync::Arc,
+};
+
+pub(crate) struct BiBTreeMap<K, V> {
+    left2right: BTreeMap<Arc<K>, Arc<V>>,
+    right2left: BTreeMap<Arc<V>, Arc<K>>,
+}
+
+impl<K, V> BiBTreeMap<K, V>
+where
+    K: std::hash::Hash + Eq + Ord + std::fmt::Display,
+    V: std::hash::Hash + Eq + Ord,
+{
+    pub(crate) fn new() -> Self {
+        BiBTreeMap {
+            left2right: BTreeMap::new(),
+            right2left: BTreeMap::new(),
+        }
+    }
+
+    pub(crate) fn insert(&mut self, key: K, value: V) {
+        let rc_key = Arc::new(key);
+        let rc_value = Arc::new(value);
+        self.right2left.insert(rc_value.clone(), rc_key.clone());
+        self.left2right.insert(rc_key, rc_value);
+    }
+
+    pub(crate) fn get_by_left(&self, key: &K) -> Option<&Arc<V>> {
+        self.left2right.get(key)
+    }
```

Comment:
You can probably use the following to directly return a reference on `V` (no `Arc` involved):

```rust
        self.left2right.get(key).map(|x| x.as_ref())
```

Same for `get_by_right`.

Reply:
Thanks, done!